### PR TITLE
Removed params from proxy's SendAsync

### PR DESCRIPTION
**Pull request by JamesNK**
https://github.com/aspnet/SignalR/issues/1268

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 namespace Microsoft.AspNetCore.SignalR
 {
-    public class DefaultHubLifetimeManager<THub> : HubLifetimeManager<THub>
+    public class DefaultHubLifetimeManager<THub> : HubLifetimeManager<THub> where THub : Hub
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs]: I noticed there were some classes with generic arguments for a hub but had no type constraint. I've added them but was it intentional? If so then let me know and I'll remove and add comments with a reason

File: src/Microsoft.AspNetCore.SignalR.Core/Proxies.cs [hunk 2]
```diff
@@ -17,13 +17,13 @@ public UserProxy(HubLifetimeManager<THub> lifetimeManager, string userId)
             _userId = userId;
         }
 
-        public Task SendAsync(string method, params object[] args)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Core/Proxies.cs]: I noticed that when the reference was `IClientProxy` then a byte array would always be the first argument but with these typed proxies the `params` would take precedence. Just removed `params` on all the `IClientProxy` implementations.

Note that `client.SendAsync((object[])bytes);` will still expand out but that seems like a strange thing to do.

File: src/Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs (hunk 1 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs]: That's good, we just forgot them.

**Comment by davidfowl** [comment]
I only have one question, should we change this to IReadOnlyList or is that an endless rathole?

**Comment by JamesNK** [comment]
@BrennanConroy Can you verify whether this fixes the original issue or not?

**Comment by BrennanConroy** [comment]
I thought there was a problem with this approach. I'll take a quick look and see if I can find any issues

**Comment by BrennanConroy** [comment]
Oh wait, if you pass `new string[] { "test", "this", "function" }` into `SendAsync` then you get 3 argument values on the client instead of one `"arguments":["test","this","function"]` but it should be a single array argument on the client?

**Comment by JamesNK** [comment]
Hmmm, so a struct array doesn't choose the `params` overload but a reference type array does. Weird behavior.

We could have `IClientProxy.SendCoreAsync(string name, object[] args)` and then all the extension methods are called `SendAsync`.

**Comment by davidfowl** [comment]
Yes, lets do the rename for the array overload. SendArrayAsync?

**Comment by JamesNK** [comment]
It isn't sending a value that is an array, it is sending an array of arguments.

I like `SendCoreAsync` since it is the core method that is doing the sending that all the other overloads call.

**Comment by davidfowl** [comment]
> I like SendCoreAsync since it is the core method that is doing the sending that all the other overloads call.

Usually, the *Core* method is an internal/private method that all other public methods delegate to. It's rarely also itself a public method and that's why I hesitate.

**Comment by JamesNK** [comment]
True, but other than being public it fulfills the rest of the core *vibe*. It is the method that isn't intended to be called by most users and is delegated to by all the `SendAsync` overloads.

**Comment by davidfowl** [comment]
Lets just rename it then. It should be rare enough.

**Comment by JamesNK** [comment]
Ok

How is this for a crazy idea though:
1. Still have `IClientProxy.SendAsyncCore(string methodName, object[] args)`
2. Abstract base class called `ClientProxy` that explicitly implements `IClientProxy.CoreSendAsync` so it isn't public
3. Instead of exposing `IClientProxy` everywhere we expose `ClientProxy`
4. `IClientProxy` extension methods should still work and because `CoreSendAsync` was explicitly implemented then it isn't visible.

Personally I don't mind SendAsyncCore being public but if you don't want it visible in intellisense...

**Comment by anurse** [comment]
Honestly I think this is one of the rare appropriate cases for `EditorBrowsableAttribute`... I do like your crazy idea, but I think just hiding it from IntelliSense would be sufficient rather than actually trying to hide it (plus there may be legitimate reasons for users to want to have access to `SendAsyncCore`)

**Comment by davidfowl** [comment]
I'd rather expose SendCoreAsync. I'm not *that* impractical.

**Pull request merged**